### PR TITLE
FP8 support for tokamax gmm kernel

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This PR calls tokamax_api.gmm in megablox/ops.py for supporting FP8 quantization using tokamax's gmm kernel. It also extends the tiling configs to 18 values, 9 for wi-* and 9 for wo tensors.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/458461066

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Command to repro: https://paste.googleplex.com/6669707558125568
Xprof: http://shortn/_HoU0Xj2IZ9

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -785,7 +785,7 @@ def sparse_matmul(
     """Perform sparse matrix multiplication of inputs and Experts."""
 
     def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
-      pad_length = self.config.tile_batch_seq
+      pad_length = self.config.wi_tile_fwd_batch_seq
```
Comment (suexu1025) [inline src/MaxText/layers/moe.py:788]: @khatwanimohit does that mean bf16 also needs to provide a validate `wi_tile_fwd_batch_seq`?

**Review by suexu1025 — APPROVED** [review]

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py:788]: yes, just like we required the right `tile_batch_seq`

File: src/MaxText/configs/base.yml [hunk 2]
```diff
@@ -178,9 +178,25 @@ use_random_routing: False # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: True # whether to use a custom sort vjp for sparse matmul ops
 use_ring_of_experts: False # whether to use ring of experts for sparse matmul expert parallelism
 # Tunable tiling dimensions used for Megablox
```
Comment (RissyRan) [inline src/MaxText/configs/base.yml]: Let's update it to something like `Tunable tiling dimensions used for MLP GMM`? which includes Tokamax gmm/ragged_dot, megablox, JAX ragged_dot?

File: src/MaxText/configs/types.py [hunk 3]
```diff
@@ -543,9 +543,24 @@ class MoEKernels(BaseModel):
 
   megablox: bool = Field(True, description="Whether to use Megablox kernels for MoE.")
   sparse_matmul: bool = Field(True, description="Whether to use sparse matmul kernels for MoE.")
-  tile_batch_seq: int = Field(512, description="Tunable tiling dimension for batch/sequence in Megablox.")
-  tile_embed_dim: int = Field(1024, description="Tunable tiling dimension for embedding in Megablox.")
-  tile_mlp_dim: int = Field(1024, description="Tunable tiling dimension for MLP in Megablox.")
+  wi_tile_fwd_batch_seq: int = Field(512, description="Tunable tiling dimension for batch/sequence in Megablox.")
```
Comment (RissyRan) [inline src/MaxText/configs/types.py]: Nit: `in Megablox` --> `in GMM` or something similar?

File: src/MaxText/configs/types.py [hunk 4]
```diff
@@ -1389,7 +1404,7 @@ class DerivedValues(BaseModel):
       None, description="Increment for global batch size during rampup."
   )
   rampup_samples_per_increment_to_load: None | float = Field(None, description="Samples per increment for rampup.")
-  tile_fwd_batch_seq: None | int = Field(None, description="Legacy alias for tile_batch_seq.")
+  # tile_fwd_batch_seq: None | int = Field(None, description="Legacy alias for tile_batch_seq.")
```
Comment (RissyRan) [inline src/MaxText/configs/types.py]: Shall we remove it if comment it out?

File: src/MaxText/configs/types.py [hunk 5]
```diff
@@ -1710,7 +1725,7 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
     if self.expert_shard_attention_option == "context":
       cp_size *= self.ici_expert_parallelism * self.dcn_expert_parallelism
     self.context_parallel_size = cp_size
-    self.tile_fwd_batch_seq = self.tile_batch_seq  # Legacy alias.
+    # self.tile_fwd_batch_seq = self.tile_batch_seq  # Legacy alias.
```
Comment (RissyRan) [inline src/MaxText/configs/types.py]: Similar comment.

File: src/MaxText/kernels/megablox/ops.py [hunk 6]
```diff
@@ -114,29 +120,47 @@ def _gmm_fwd(
...
-      transpose_rhs=transpose_rhs,
-      interpret=interpret,
-  )
+      # QAG is only supported for following conditions
+  if use_tokamax_backend:
+    if quantization_rule and quantization_rule.bwd_qtype:
+      if quantization_rule.weight_calibration_method.startswith("fixed") and isinstance(rhs, qpl.QArray):
+        rhs_qvalue = jax.lax.all_gather(rhs.qvalue, "fsdp", axis=0, tiled=True)
+        rhs = dataclasses.replace(rhs, qvalue=rhs_qvalue)
+    out = tokamax_backend.gmm(
```
Comment (RissyRan) [inline src/MaxText/kernels/megablox/ops.py:135]: Ah... It looks like we integrated the tokamax quantization logic into megablox ops.

File: src/MaxText/configs/base.yml (hunk 2 above)
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: jax.lax.ragged_dot just accepts 3 tilings configs, I'll update the comment for others

File: src/MaxText/kernels/megablox/ops.py [hunk 7]
```diff
@@ -187,27 +210,52 @@ def _gmm_bwd(
...
+        group_sizes=group_sizes,
+        precision=jax.lax.Precision.DEFAULT,
+        out_dtype=rhs_dtype,
+        tiling=tiling[-3:],
+        group_offset=group_offset,
+        num_actual_groups=num_actual_groups,
+        interpret=interpret,
+    )
+    if quantization_rule and quantization_rule.bwd_qtype:
+      drhs = jax.lax.psum_scatter(drhs, "fsdp", scatter_dimension=0, tiled=True)
```
Comment (RissyRan) [inline src/MaxText/kernels/megablox/ops.py:245]: Just curious the reason we need RS for fsdp :) It usually happens in TP. This is on purpose I guess?

File: src/MaxText/layers/moe.py [hunk 8]
```diff
@@ -804,20 +804,46 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
+            tiling[7],
+            tiling[8],
         )
+      else:
+        tiling = (
+            min(tiling[0], m),
+            min(tiling[1], k),
+            min(tiling[2], n),
+        )
+      if self.config.use_tokamax_gmm:
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py:813]: Can we put tiling logic here inside of use_tokamax_gmm?

```
      if self.config.megablox:
        tiling = (
            min(tiling[0], m),
            min(tiling[1], k),
            min(tiling[2], n),
            tiling[3],
            tiling[4],
            tiling[5],
            tiling[6],
            tiling[7],
            tiling[8],
        )
```

File: src/MaxText/layers/moe.py [hunk 9]
```diff
@@ -903,15 +930,22 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
     # w0, w1, wo needs to be un sharded on fsdp / fsdp_transpose axis, so use
     # mlp_no_fsdp axis
-    if self.config.fsdp_shard_on_exp:
-      # special sharding for dsv3 to remove overhead between gmm/AG
-      w0_pspec = nn.logical_to_mesh_axes(("embed_tensor_transpose", None, "mlp_no_fsdp"))
-      w1_pspec = nn.logical_to_mesh_axes(("embed_tensor_transpose", None, "mlp_no_fsdp"))
-      wo_pspec = nn.logical_to_mesh_axes(("embed_tensor_transpose", "mlp_no_fsdp", None))
+    if self.config.quantization:
+      if not self.config.fsdp_shard_on_exp:
+        raise ValueError("Quantization is only supported with fsdp_shard_on_exp=True.")
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Can we move this assertion to pyconfig.py to early return?

File: src/MaxText/configs/types.py [hunk 10]
```diff
@@ -543,9 +543,24 @@ class MoEKernels(BaseModel):
...
+  wi_tile_drhs_embed_dim: int = Field(1024, description="Tunable tiling dimension for embedding in Megablox.")
+  wi_tile_drhs_mlp_dim: int = Field(1024, description="Tunable tiling dimension for MLP in Megablox.")
+  wo_tile_fwd_batch_seq: int = Field(512, description="Tunable tiling dimension for batch/sequence in Megablox.")
+  wo_tile_fwd_embed_dim: int = Field(1024, description="Tunable tiling dimension for embedding in Megablox.")
+  wo_tile_fwd_mlp_dim: int = Field(1024, description="Tunable tiling dimension for MLP in Megablox.")
+  wo_tile_dlhs_batch_seq: int = Field(512, description="Tunable tiling dimension for batch/sequence in Megablox.")
+  wo_tile_dlhs_embed_dim: int = Field(1024, description="Tunable tiling dimension for embedding in Megablox.")
+  wo_tile_dlhs_mlp_dim: int = Field(1024, description="Tunable tiling dimension for MLP in Megablox.")
+  wo_tile_drhs_batch_seq: int = Field(512, description="Tunable tiling dimension for batch/sequence in Megablox.")
+  wo_tile_drhs_embed_dim: int = Field(1024, description="Tunable tiling dimension for embedding in Megablox.")
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply src/MaxText/configs/types.py]: updated

File: src/MaxText/kernels/megablox/ops.py [hunk 11]
```diff
@@ -146,7 +170,6 @@ def _gmm_bwd(
     grad: jnp.ndarray,
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply src/MaxText/kernels/megablox/ops.py:177]: added back the del. we need to keep signature aligned with the fwd pass

File: src/MaxText/configs/types.py (hunk 3 above)
Comment (khatwanimohit) [reply src/MaxText/configs/types.py]: done

File: src/MaxText/configs/types.py (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/configs/types.py]: done

File: src/MaxText/layers/moe.py (hunk 8 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py:813]: modified the condition

**Review by RissyRan — APPROVED** [review]
LGTM! Just one comment to move proper assertion in moe.py to the pyconfig file

File: src/MaxText/layers/moe.py (hunk 9 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py]: I have now removed the assertion and made changes in ops.py so things work with quantization and fsdp_shard_on_exp=false

File: src/MaxText/layers/moe.py [hunk 12]
```diff
@@ -804,20 +804,47 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
+            min(tiling[0], m),
+            min(tiling[1], k),
+            min(tiling[2], n),
+            tiling[3],
+            tiling[4],
+            tiling[5],
+            tiling[6],
+            tiling[7],
+            tiling[8],
+        )
```
Comment (BirdsOfAFthr) [inline src/MaxText/layers/moe.py]: Please completely remove these

File: src/MaxText/kernels/megablox/backend.py
Comment (BirdsOfAFthr) [inline src/MaxText/kernels/megablox/backend.py:1]: nit: there is no need to update these in backend. I was just using these code to debug

Comment (khatwanimohit) [reply src/MaxText/kernels/megablox/backend.py:1]: I like the fact that we can see the tiling in xprof if ever we want to use the megablox path

File: src/MaxText/layers/moe.py (hunk 12 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py]: removed

File: src/MaxText/configs/base.yml
Comment (BirdsOfAFthr) [inline src/MaxText/configs/base.yml:1]: Should we add the fine-grained calibration method? My recipe only applies static scaling to weights and activations which I hardcoded there.

File: src/MaxText/kernels/megablox/backend.py
Comment (BirdsOfAFthr) [reply src/MaxText/kernels/megablox/backend.py:1]: ok, but I though we're removing this backend.py. It wont be used anywhere

File: src/MaxText/configs/base.yml
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml:1]: exposed fine-grained options for calibration method
